eventmetrics/parsers/timeElapsedParser.go

```diff
@@ -0,0 +1,224 @@
+	firmwareMetadataKey     = "/fw-name"
+	rebootReasonMetadataKey = "/hw-last-reboot-reason"
+)
+
+// EventClient is an interface that provides a list of events related to a device.
+type EventClient interface {
+	GetEvents(deviceID string) []interpreter.Event
+}
+
+// Finder returns a specific event in a list of events.
+type Finder interface {
+	Find(events []interpreter.Event, incomingEvent interpreter.Event) (interpreter.Event, error)
+}
+
+// ErrorWithEvent is an optional interface that errors can implement if the error contains an event that is related
+// to the error.
+type ErrorWithEvent interface {
+	Event() interpreter.Event
+}
+
+// TimeElapsedConfig holds the configuration for a time-elapsed parser.
+type TimeElapsedConfig struct {
+	Name            string
+	SearchedSession string
+	IncomingEvent   EventConfig
+	SearchedEvent   EventConfig
+}
+
+// TimeElapsedParser is a parser that calculates the time between two events.
+type TimeElapsedParser struct {
+	finder        Finder
+	incomingEvent EventInfo
+	searchedEvent EventInfo
+	name          string
+	logger        log.Logger
+	client        EventClient
+	measures      Measures
+}
+
+// NewTimeElapsedParser creates a TimeElapsedParser from a TimeElapsedConfig or returns an error if there one cannot be created.
+func NewTimeElapsedParser(config TimeElapsedConfig, client EventClient, logger log.Logger, measures Measures, currentTime TimeFunc) (*TimeElapsedParser, error) {
+	incomingEvent, err := NewEventInfo(config.IncomingEvent, currentTime)
+	if err != nil {
+		return nil, err
+	}
+
+	var searchedEvent EventInfo
+	if len(config.SearchedEvent.Regex) == 0 {
+		searchedEvent = EventInfo{Regex: incomingEvent.Regex, CalculateUsing: Boottime}
+		incomingEvent.CalculateUsing = Birthdate
+	} else {
+		searchedEvent, err = NewEventInfo(config.SearchedEvent, currentTime)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var finder Finder
+	sessionType := ParseSessionType(config.SearchedSession)
+	comparators := history.Comparators([]history.Comparator{
+		history.OlderBootTimeComparator(),
+		history.DuplicateEventComparator(incomingEvent.Regex),
+	})
+
+	if len(config.SearchedEvent.Regex) == 0 {
+		finder = history.EventHistoryIterator(comparators)
+	} else if sessionType == Current {
+		finder = history.CurrentSessionFinder(searchedEvent.Validator, comparators)
+	} else {
+		finder = history.LastSessionFinder(searchedEvent.Validator, comparators)
+	}
+
+	return &TimeElapsedParser{
+		finder:        finder,
+		incomingEvent: incomingEvent,
+		searchedEvent: searchedEvent,
+		logger:        logger,
+		client:        client,
+		measures:      measures,
+		name:          config.Name,
+	}, nil
+}
+
+// Parse implements the Parser interface. Calculates the time-elapsed and then adds to prometheus metrics if possible.
+func (t *TimeElapsedParser) Parse(event interpreter.Event) {
+	restartTime, err := t.calculateTimeElapsed(event)
+	if err != nil || restartTime <= 0 {
+		return
```

Comment:
Do we want this case to be counted for the unparseable events metric?

Reply:
Good point! Would be helpful for monitoring/debugging any problems.